### PR TITLE
[Logs]: Added support of docker labels to enrich logs

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

- Implemented a logic to extract logs config attached to containers labels.
- Isolated the code to initialise the docker client
- Isolated the logic to filter in containers to tail
- Added unit-tests

### Motivation

Be able to tail containers that have a logs config attached to their labels.

### CODE REVIEW

File: pkg/logs/input/container/docker.go [hunk 1]
```diff
@@ -198,7 +211,13 @@ func (dt *DockerTailer) forwardMessages() {
 		}
 		containerMsg := message.NewContainerMessage(updatedMsg)
 		msgOrigin := message.NewOrigin()
-		msgOrigin.LogSource = dt.source
+		// use logs metada instead of source config when defined,
+		// it might interesting to merge both at some point.
+		if dt.logsMetadata != nil {
+			msgOrigin.LogsConfig = dt.source.Config
+		} else {
+			msgOrigin.LogsConfig = dt.logsMetadata
```
Comment (tmichelet) [inline pkg/logs/input/container/docker.go]: should it be the opposite?

File: pkg/logs/input/container/docker.go [hunk 2]
```diff
@@ -208,6 +227,34 @@ func (dt *DockerTailer) forwardMessages() {
...
+	label, exists := information.Config.Labels[logsMetadataPath]
+	if !exists {
+		return nil, false
+	}
+	return dt.computeLogsMetadataFromLabel(label)
+}
+
+// computeLogsMetadataFromLabel returns logs metadata if label is correctly formatted.
+func (dt *DockerTailer) computeLogsMetadataFromLabel(label string) (*config.LogsConfig, bool) {
+	var configs []config.LogsConfig
```
Comment (tmichelet) [inline pkg/logs/input/container/docker.go]: do we want to have a `[]config.LogConfig` or simply a `config.LogConfig`?

Comment (ajacquemot) [reply pkg/logs/input/container/docker.go]: we expect the label to be like:
```
LABEL "com.datadoghq.ad.logs"='[{"source": "any_source", "service": "any_service"}]'
```

File: pkg/logs/input/container/docker.go (hunk 1 above)
Comment (ajacquemot) [reply pkg/logs/input/container/docker.go]: Good catch

File: pkg/logs/input/container/docker.go (hunk 2 above)
Comment (ajacquemot) [reply pkg/logs/input/container/docker.go]: It's worthy to add a description though

File: pkg/logs/input/container/docker.go [hunk 3]
```diff
@@ -208,6 +227,35 @@ func (dt *DockerTailer) forwardMessages() {
...
+// and its value respects format '[{...}]'.
+func (dt *DockerTailer) extractLogsMetadataFromLabels(labels map[string]string) (*config.LogsConfig, bool) {
+	label, exists := labels[logsMetadataPath]
+	if !exists {
+		return nil, false
+	}
+	var configs []config.LogsConfig
+	err := json.Unmarshal([]byte(label), &configs)
+	if err != nil {
+		return nil, false
```
Comment (mnshdw) [inline pkg/logs/input/container/docker.go]: It might be interesting to log we tried but did not succeed in reading the labels into proper metadata here.

File: pkg/logs/input/container/docker_test.go [hunk 4]
```diff
@@ -81,7 +81,48 @@ func (suite *DockerTailerTestSuite) TestParseMessage() {
...
+func (suite *DockerTailerTestSuite) TestComputeLogsMetadataFromLabel() {
+	var labels map[string]string
+	var config *config.LogsConfig
+	var valid bool
+
+	config, valid = suite.tailer.extractLogsMetadataFromLabels(labels)
+	suite.Nil(config)
+	suite.False(valid)
+
+	// logs metadata should be null because of wrong format
```
Comment (mnshdw) [inline pkg/logs/input/container/docker_test.go]: A common practice is to do smaller unit tests and name them after what you're testing (just like what you are doing in the comment).

File: pkg/logs/input/container/docker.go [hunk 5]
```diff
@@ -198,7 +211,13 @@ func (dt *DockerTailer) forwardMessages() {
 		}
 		containerMsg := message.NewContainerMessage(updatedMsg)
 		msgOrigin := message.NewOrigin()
-		msgOrigin.LogSource = dt.source
```
Comment (mnshdw) [inline pkg/logs/input/container/docker.go]: I think it's a good move to store only the config and not whole source, I just have one question: are we sure we won't need the source elsewhere downstream ?

File: pkg/logs/input/container/docker.go [hunk 6]
```diff
@@ -208,6 +227,35 @@ func (dt *DockerTailer) forwardMessages() {
 	}
 }
 
+// fetchLogsMetadata returns the logs metedata defined in container labels.
```
Comment (mnshdw) [inline pkg/logs/input/container/docker.go]: typo `s/metedata/metadata`

File: pkg/logs/input/container/docker.go [hunk 7]
```diff
@@ -198,7 +211,13 @@ func (dt *DockerTailer) forwardMessages() {
 		}
 		containerMsg := message.NewContainerMessage(updatedMsg)
 		msgOrigin := message.NewOrigin()
-		msgOrigin.LogSource = dt.source
+		// use logs metada instead of source config when defined,
```
Comment (mnshdw) [inline pkg/logs/input/container/docker.go]: typo `s/metada/metadata`

File: pkg/logs/input/container/docker.go [hunk 8]
```diff
@@ -42,6 +47,7 @@ type DockerTailer struct {
 	cli           *client.Client
 	source        *config.LogSource
 	containerTags []string
+	logsMetadata  *config.LogsConfig
```
Comment (mnshdw) [inline pkg/logs/input/container/docker.go]: It seems like an overkill to me. You don't want to embed the whole LogsConfig if you only use `source` and `service`. Could we have a dedicated struct ?

File: pkg/logs/input/container/docker.go [hunk 9]
```diff
@@ -208,6 +227,35 @@ func (dt *DockerTailer) forwardMessages() {
...
+
+// extractLogsMetadataFromLabels returns logs metadata extracted from labels when it contains key 'com.datadoghq.ad.logs'
+// and its value respects format '[{...}]'.
+func (dt *DockerTailer) extractLogsMetadataFromLabels(labels map[string]string) (*config.LogsConfig, bool) {
+	label, exists := labels[logsMetadataPath]
+	if !exists {
+		return nil, false
+	}
+	var configs []config.LogsConfig
+	err := json.Unmarshal([]byte(label), &configs)
```
Comment (mnshdw) [inline pkg/logs/input/container/docker.go]: Another reason why you don't want a `LogsConfig`, you don't want parsing to fail for unrelated stuff.

**Review by mnshdw — APPROVED** [review]
Looks good, left a few comments that you might want to address, there's only one thing that I'd require to change: we probably don't want to use the whole `LogsConfig` to store two fields.

**Review by mnshdw — CHANGES_REQUESTED** [review]
Picked the wrong status ^^

File: pkg/logs/input/container/docker.go (hunk 8 above)
Comment (ajacquemot) [reply pkg/logs/input/container/docker.go]: I am not sure about this, I think a user could enable the `multiline` from here, WDYT ? Maybe @NBParis has a strong opinion

File: pkg/logs/input/container/docker.go (hunk 5 above)
Comment (ajacquemot) [reply pkg/logs/input/container/docker.go]: As far as we don't need to report any status downstream, we should be good.

File: pkg/logs/input/container/docker.go (hunk 8 above)
Comment (NBParis) [reply pkg/logs/input/container/docker.go]: We need to be able to specified processing rule here include, exclude, multiline and scrubbing.

**Review by mnshdw — DISMISSED** [review]

File: pkg/logs/input/container/client.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	if err != nil {
+		return nil, err
+	}
+	client.UpdateClientVersion(clientVersion)
+	return client, nil
+}
+
+// serverAPIVersion returns the latest version of the docker API supported by the docker server
+func getServerAPIVersion(client *client.Client) (string, error) {
+	client.UpdateClientVersion("")
```
Comment (mnshdw) [inline pkg/logs/input/container/client.go]: What does `UpdateClientVersion("")` and why do we need it ?

File: pkg/logs/input/container/client.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	v, err := client.ServerVersion(context.Background())
+	if err != nil {
+		return "", err
+	}
+	return v.APIVersion, nil
+}
+
+// computeAPIVersion returns the version of the API that the docker client should use to be able to communicate with the server
+func computeClientAPIVersion(apiVersion string) (string, error) {
```
Comment (mnshdw) [inline pkg/logs/input/container/client.go]: So `apiVersion` is in fact `serverVersion`, why not name it this way for clarity, or better yet directly pass the docker client as an argument here ?

File: pkg/logs/input/container/filter.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+		if source := searchSource(container, sources); source != nil {
+			containersToTail = append(containersToTail, NewContainer(container, source))
+		}
+	}
+	return containersToTail
+}
+
+// searchSource returns a valid source for container,
+// if no source is found, return nil
+func searchSource(container types.Container, sources []*config.LogSource) *config.LogSource {
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: What prevents you from defining this func directly on a Container ?

File: pkg/logs/input/container/filter.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+func NewContainer(container types.Container, source *config.LogSource) *Container {
+	return &Container{
+		Identifier: container.ID,
+		Image:      container.Image,
+		Source:     source,
+	}
+}
+
+// Filter returns all the containers for which a source exists.
+func Filter(containers []types.Container, sources []*config.LogSource) []*Container {
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: It sounds like this is the responsibility of the scanner.

File: pkg/logs/input/container/filter.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		if computeScore(container, candidate) < computeScore(container, source) {
+			candidate = source
+		}
+	}
+	return candidate
+}
+
+// computeScore returns the matching score between the container and the source,
+// the higher, the better
+func computeScore(container types.Container, source *config.LogSource) int {
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: Same as above: What prevents you from defining this func directly on a Container ?

File: pkg/logs/input/container/filter.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+	if isLabelMatch(source.Config.Label, container.Labels) {
+		score++
+	}
+	return score
+}
+
+// digestPrefix represents a prefix that can be added to an image name.
+const digestPrefix = "@sha256:"
+
+// isImageMatch returns true if image respects format: "Y/imageFilter@sha256:X" with 'Y/' and '@sha256:X' optional.
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: I don't really understand the format nor the variables here. Can we try to use a clearer doc for explaining the possible image formats ? eg `[<repository>/]image[@sha256:<digest>]` and name the function variables accordingly.

File: pkg/logs/input/container/filter.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+// the config has to be conform with the format '[{...}]'.
+func extractLogsConfig(labels map[string]string) *config.LogsConfig {
+	label, exists := labels[logsConfigPath]
+	if !exists {
+		return nil
+	}
+	var configs []config.LogsConfig
+	err := json.Unmarshal([]byte(label), &configs)
+	if err != nil || len(configs) < 1 {
+		log.Warnf("Could not parse logs configs, got %v, expect value with format '[{\"source\":\"a_source\",\"service\":\"a_service\", ...}]'")
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: Not necessarily a big problem, but I am a bit wary of warning logs that act as documentation. I'm not sure if we modify the source format we will always think to update this line.

File: pkg/logs/input/container/filter.go [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+	containersToTail := []*Container{}
+	for _, container := range containers {
+		if source := searchSource(container, sources); source != nil {
+			containersToTail = append(containersToTail, NewContainer(container, source))
+		}
+	}
+	return containersToTail
+}
+
+// searchSource returns a valid source for container,
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: Maybe document it's rather the source that most closely matches the container.

File: pkg/logs/input/container/filter.go [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+		if _, exists := labels[label]; exists || len(parts) == 2 && labels[parts[0]] == parts[1] {
+			return true
+		}
+	}
+	return false
+}
+
+// logsConfigPath refers to the logs configuration that can be passed over a docker label,
+// this feature is commonly named autodicovery.
+const logsConfigPath = "com.datadoghq.ad.logs"
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: What's `ad` ?

File: pkg/logs/input/container/filter.go [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+// sourceFromContainer returns the source extracted and computed from the container labels.
+func sourceFromContainer(container types.Container) *config.LogSource {
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: nit: `(c *types.Container) toSource() *config.LogSource {` ?

File: pkg/logs/input/container/filter.go [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+	logsConfig := extractLogsConfig(container.Labels)
+	if logsConfig == nil {
+		return nil
+	}
+	return config.NewLogSource(logsConfigPath, logsConfig)
+}
+
+// extractLogsConfig returns the logs config present in the label 'com.datadoghq.ad.logs',
+// the config has to be conform with the format '[{...}]'.
+func extractLogsConfig(labels map[string]string) *config.LogsConfig {
```
Comment (mnshdw) [inline pkg/logs/input/container/filter.go]: I think you could improve the quality of your code by refining the encapsulation of your variables.
I'll take this one as an example but there are others around:
- `extractLogsConfig(labels map[string]string)` lacks cohesion because the caller leaks `.Labels`. How is a label map related to a config ? It should be an implementation detail not visible from the method signature.
- `extractLogsConfig(container types.Container)` is better (more self-contained).
The method name could still be improved.

By the way, I don't think we need to repeat `Logs` all over the place, we already are in a `logs` pkg.

**Review by mnshdw — COMMENTED** [review]
Overall I like the direction this is going.

I left quite a few comments, there's nothing dramatic but I feel like we could really improve the readability of the code.

File: pkg/logs/input/container/container.go [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+		score++
+	}
+	return score
+}
+
+// digestPrefix represents a prefix that can be added to an image name.
+const digestPrefix = "@sha256:"
+
+// isImageMatch returns true if container image respects format: '[<repository>/]image[@sha256:<digest>]'.
+func (c *Container) isImageMatch(imageFilter string) bool {
```
Comment (mnshdw) [inline pkg/logs/input/container/container.go]: What is `imageFilter` in the format ?

**Review by mnshdw — APPROVED** [review]
